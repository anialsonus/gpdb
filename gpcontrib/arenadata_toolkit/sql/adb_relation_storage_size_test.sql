CREATE EXTENSION arenadata_toolkit;

CREATE TABLE heap_table_with_toast(a INT, b TEXT)
DISTRIBUTED BY (a);

CREATE TABLE heap_table_without_toast(a INT, b INT)
DISTRIBUTED BY (a);

CREATE TABLE ao_table_with_toast(a INT, b TEXT)
WITH (APPENDOPTIMIZED=true, ORIENTATION=COLUMN)
DISTRIBUTED BY (a);

CREATE TABLE ao_table_without_toast(a INT, b INT)
WITH (APPENDOPTIMIZED=true, ORIENTATION=COLUMN)
DISTRIBUTED BY (a);

-- Check that toast exist only for "with_toast" tables
SELECT relname
FROM pg_class
WHERE reltoastrelid = 0 AND
      relnamespace IN (SELECT oid FROM pg_namespace WHERE nspname = 'public');

SELECT relname
FROM pg_class
WHERE reltoastrelid != 0 AND
      relnamespace IN (SELECT oid FROM pg_namespace WHERE nspname = 'public');

-- Insert initial data to tables
INSERT INTO heap_table_with_toast SELECT i, 'short_text' FROM generate_series(1,15) AS i;
INSERT INTO heap_table_without_toast SELECT i, i*10 FROM generate_series(1,15) AS i;
INSERT INTO ao_table_with_toast SELECT i, 'short_text' FROM generate_series(1,15) AS i;
INSERT INTO ao_table_without_toast SELECT i, i*10 FROM generate_series(1,15) AS i;

-- Check sizes on segments
SELECT * FROM arenadata_toolkit.adb_relation_storage_size_on_segments((
    SELECT oid FROM pg_class WHERE relname = 'heap_table_with_toast'))
ORDER BY 1;

SELECT * FROM arenadata_toolkit.adb_relation_storage_size_on_segments((
    SELECT oid FROM pg_class WHERE relname = 'heap_table_without_toast'))
ORDER BY 1;

SELECT * FROM arenadata_toolkit.adb_relation_storage_size_on_segments((
    SELECT oid FROM pg_class WHERE relname = 'ao_table_with_toast'))
ORDER BY 1;

SELECT * FROM arenadata_toolkit.adb_relation_storage_size_on_segments((
    SELECT oid FROM pg_class WHERE relname = 'ao_table_without_toast'))
ORDER BY 1;

-- Add random large data to get non-zero toast table's size
UPDATE heap_table_with_toast SET b = (
    SELECT string_agg( chr(trunc(65+random()*26)::integer), '')
    FROM generate_series(1,50000))
WHERE a = 1;

UPDATE ao_table_with_toast SET b = (
    SELECT string_agg( chr(trunc(65+random()*26)::integer), '')
    FROM generate_series(1,50000))
WHERE a = 1;

SELECT * FROM arenadata_toolkit.adb_relation_storage_size_on_segments((
    SELECT oid FROM pg_class WHERE relname = 'heap_table_with_toast'))
ORDER BY 1;

SELECT * FROM arenadata_toolkit.adb_relation_storage_size_on_segments((
    SELECT oid FROM pg_class WHERE relname = 'ao_table_with_toast'))
ORDER BY 1;

-- Check summary size of tables
SELECT * FROM arenadata_toolkit.adb_relation_storage_size((
    SELECT oid FROM pg_class WHERE relname = 'heap_table_with_toast'));

SELECT * FROM arenadata_toolkit.adb_relation_storage_size((
    SELECT oid FROM pg_class WHERE relname = 'heap_table_without_toast'));

SELECT * FROM arenadata_toolkit.adb_relation_storage_size((
    SELECT oid FROM pg_class WHERE relname = 'ao_table_with_toast'));

SELECT * FROM arenadata_toolkit.adb_relation_storage_size((
    SELECT oid FROM pg_class WHERE relname = 'ao_table_without_toast'));

-- Cleanup
DROP TABLE heap_table_with_toast;
DROP TABLE heap_table_without_toast;
DROP TABLE ao_table_with_toast;
DROP TABLE ao_table_without_toast;

DROP EXTENSION arenadata_toolkit;
