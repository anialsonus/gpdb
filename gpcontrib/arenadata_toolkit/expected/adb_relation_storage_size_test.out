CREATE EXTENSION arenadata_toolkit;
CREATE TABLE heap_table_with_toast(a INT, b TEXT)
DISTRIBUTED BY (a);
CREATE TABLE heap_table_without_toast(a INT, b INT)
DISTRIBUTED BY (a);
CREATE TABLE ao_table_with_toast(a INT, b TEXT)
WITH (APPENDOPTIMIZED=true, ORIENTATION=COLUMN)
DISTRIBUTED BY (a);
CREATE TABLE ao_table_without_toast(a INT, b INT)
WITH (APPENDOPTIMIZED=true, ORIENTATION=COLUMN)
DISTRIBUTED BY (a);
-- Check that toast exist only for first table
SELECT relname
FROM pg_class
WHERE reltoastrelid = 0 AND
      relnamespace IN (SELECT oid FROM pg_namespace WHERE nspname = 'public');
         relname          
--------------------------
 heap_table_without_toast
 ao_table_without_toast
(2 rows)

SELECT relname
FROM pg_class
WHERE reltoastrelid != 0 AND
      relnamespace IN (SELECT oid FROM pg_namespace WHERE nspname = 'public');
        relname        
-----------------------
 heap_table_with_toast
 ao_table_with_toast
(2 rows)

-- Insert initial data to tables
INSERT INTO heap_table_with_toast SELECT i, 'short_text' FROM generate_series(1,15) AS i;
INSERT INTO heap_table_without_toast SELECT i, i*10 FROM generate_series(1,15) AS i;
INSERT INTO ao_table_with_toast SELECT i, 'short_text' FROM generate_series(1,15) AS i;
INSERT INTO ao_table_without_toast SELECT i, i*10 FROM generate_series(1,15) AS i;
-- Check sizes on segments
SELECT * FROM arenadata_toolkit.adb_relation_storage_size_on_segments((
    SELECT oid FROM pg_class WHERE relname = 'heap_table_with_toast'))
ORDER BY 1;
 gp_segment_id | size  
---------------+-------
             0 | 32768
             1 | 32768
             2 | 32768
(3 rows)

SELECT * FROM arenadata_toolkit.adb_relation_storage_size_on_segments((
    SELECT oid FROM pg_class WHERE relname = 'heap_table_without_toast'))
ORDER BY 1;
 gp_segment_id | size  
---------------+-------
             0 | 32768
             1 | 32768
             2 | 32768
(3 rows)

SELECT * FROM arenadata_toolkit.adb_relation_storage_size_on_segments((
    SELECT oid FROM pg_class WHERE relname = 'ao_table_with_toast'))
ORDER BY 1;
 gp_segment_id | size 
---------------+------
             0 |  160
             1 |  136
             2 |  192
(3 rows)

SELECT * FROM arenadata_toolkit.adb_relation_storage_size_on_segments((
    SELECT oid FROM pg_class WHERE relname = 'ao_table_without_toast'))
ORDER BY 1;
 gp_segment_id | size 
---------------+------
             0 |  128
             1 |  112
             2 |  144
(3 rows)

-- Add random large data to get non-zero toast table's size
UPDATE heap_table_with_toast SET b = (
    SELECT string_agg( chr(trunc(65+random()*26)::integer), '')
    FROM generate_series(1,50000))
WHERE a = 1;
UPDATE ao_table_with_toast SET b = (
    SELECT string_agg( chr(trunc(65+random()*26)::integer), '')
    FROM generate_series(1,50000))
WHERE a = 1;
SELECT * FROM arenadata_toolkit.adb_relation_storage_size_on_segments((
    SELECT oid FROM pg_class WHERE relname = 'heap_table_with_toast'))
ORDER BY 1;
 gp_segment_id | size  
---------------+-------
             0 | 32768
             1 | 98304
             2 | 32768
(3 rows)

SELECT * FROM arenadata_toolkit.adb_relation_storage_size_on_segments((
    SELECT oid FROM pg_class WHERE relname = 'ao_table_with_toast'))
ORDER BY 1;
 gp_segment_id | size  
---------------+-------
             0 |   160
             1 | 50248
             2 |   192
(3 rows)

-- Check summary size of tables
SELECT * FROM arenadata_toolkit.adb_relation_storage_size((
    SELECT oid FROM pg_class WHERE relname = 'heap_table_with_toast'));
 adb_relation_storage_size 
---------------------------
                    163840
(1 row)

SELECT * FROM arenadata_toolkit.adb_relation_storage_size((
    SELECT oid FROM pg_class WHERE relname = 'heap_table_without_toast'));
 adb_relation_storage_size 
---------------------------
                     98304
(1 row)

SELECT * FROM arenadata_toolkit.adb_relation_storage_size((
    SELECT oid FROM pg_class WHERE relname = 'ao_table_with_toast'));
 adb_relation_storage_size 
---------------------------
                     50600
(1 row)

SELECT * FROM arenadata_toolkit.adb_relation_storage_size((
    SELECT oid FROM pg_class WHERE relname = 'ao_table_without_toast'));
 adb_relation_storage_size 
---------------------------
                       384
(1 row)

-- Cleanup
DROP TABLE heap_table_with_toast;
DROP TABLE heap_table_without_toast;
DROP TABLE ao_table_with_toast;
DROP TABLE ao_table_without_toast;
DROP EXTENSION arenadata_toolkit;
